### PR TITLE
fixed the truncation operation for the POT-C & PRIT-C uniformity test. 

### LINKED ISSUE #461: Cauchy combination tests: truncation operation inconsistent with paper specification.
## What happened

1.  The `pot_c` and `prit_c` functions implement  the POT-C and PRIT-C uniformity testing procedures  of Tesso & Vehtari (["LOO-PIT predictive model checking"]( https://arxiv.org/pdf/2603.02928) ).  These combine pointwise p-values with the *(truncated) Cauchy combination method*. When using the truncated version,  the  statistic *Eq. (24)*  is the truncated Cauchy sum divided by the **total** count $n$:
$$T = \frac{1}{n} \sum_{i} \tan\left((0.5 - p_i)\pi\right) \cdot \mathbf{1}(p_i < 0.5)$$

The current code replaces this truncation operation with an **indexing-then-mean** operation, which instead divides the truncated sum by the length of the subset of p-values lower than $0.5$.

``` R
.cauchy_combination_test <- function(x, truncate = NULL) {
  if (truncate) {
    idx <- which(x < 0.5)
    if (length(idx) == 0) {
      stop("Cannot compute truncated Cauchy combination test. ",
           "No p-values below 0.5 found.")
    }
    1 - pcauchy(mean(-qcauchy(x[idx])))         # truncated sum / #retained
  } else {
    1 - pcauchy(mean(-qcauchy(x)))
  }
} 
```
With this implementation, the statistic nearly doubles under the null;  the retained count is approximately $n/2$, so the  denominator change inflates the statistic and, consequently, the rejection rate.

2.  Individual p-values from the Binomial-based PRIT-C test, computed from discrete PIT values, can exceed 1 for certain samples, which subsequently produces `NaN` values upon transformation to the Cauchy space.
``` R
.prit_test <- function(x) {
  n <- length(x)
  scaled_ecdf <- n * ecdf(x)(x)
  probs1 <- pbinom(scaled_ecdf - 1, n, x)
  probs2 <- pbinom(scaled_ecdf, n, x)
  p_values <- 2 * pmin(1 - probs1, probs2)   # can often exceed 1 

  return(p_values)
}
```

## What I expected 

1. Truncated Cauchy sum should be averaged over the total number of pointwise p-values $n$. 

``` R
.cauchy_combination_test <- function(x, truncate = NULL) {
  if (truncate){
    1 - pcauchy( mean(ifelse(x < 0.5,-qcauchy(x), 0.0)) )  # truncated sum/ n
  } else {
    1 - pcauchy(mean(-qcauchy(x)))
  }
}
```
2. Suggested fix for PRIT-C: using mid-pvalue correction can resolve the boundary issue and yield better-calibrated p-values for discrete ( rank-based) PIT.

``` R
.prit_test <- function(x) {
  n <- length(x)
  scaled_ecdf <- n * ecdf(x)(x)

  p_exact <- dbinom(scaled_ecdf,n,x)
  probs1 <- pbinom(scaled_ecdf, n, x)-0.5*p_exact
  probs2 <- 1-probs1
  
  p_values <- 2 * pmin(probs1, probs2)
  
  return(p_values)
}
```

## Environment
```
R version 4.3.3 (2024-02-29)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 24.04.3 LTS

posterior 1.6.1
```

### PR DESCRIPTION
**Pull request by hermanFTT**
Closes #461 

#### Summary

1. Truncation operation  of  POT-C and PRIT-C Cauchy combinaision  test of Tesso & Vehtari  currently implemented with  "indexing-then-mean" operation  have been replaced to align with the actual paper formulation. The mean operation should be over the total number of pointwise p-values instead of over the length of the subset of p-values greater than 0.5 ( see issue #461 ). 

1.  An adjustment for PRIT-C based on mid p-values correction is suggested to avoid p-values greater than 1. 

1. A calibration unit test is also included in the test folder. 

#### AI disclosure

No AI assistance used for this PR. 

#### Copyright and Licensing

By submitting this pull request, the copyright holder is agreeing to 
license the submitted work under the following licenses:

- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: R/uniformity_test.R [hunk 1]
```diff
@@ -166,13 +166,17 @@ uniformity_test <- function(pit, test, truncate = NULL) {
 .prit_test <- function(x) {
   n <- length(x)
   scaled_ecdf <- n * ecdf(x)(x)
-  probs1 <- pbinom(scaled_ecdf - 1, n, x)
-  probs2 <- pbinom(scaled_ecdf, n, x)
-  p_values <- 2 * pmin(1 - probs1, probs2)
 
+  p_exact <- dbinom(scaled_ecdf,n,x)
+  probs1 <- pbinom(scaled_ecdf, n, x)-0.5*p_exact
+  probs2 <- 1-probs1
```
Comment (florence-bockting) [inline R/uniformity_test.R]: Some minor changes in style. 
We use generally a white space after commas and before/after symbols (like -,+, etc.)

```suggestion
  p_exact <- dbinom(scaled_ecdf, n, x)
  probs1 <- pbinom(scaled_ecdf, n, x) - 0.5 * p_exact
  probs2 <- 1 - probs1
```

File: R/uniformity_test.R [hunk 2]
```diff
@@ -166,13 +166,17 @@ uniformity_test <- function(pit, test, truncate = NULL) {
...
+  
+  p_values <- 2 * pmin(probs1, probs2)
+  
   return(p_values)
 }
 
+
```
Comment (florence-bockting) [inline R/uniformity_test.R]: I think this extra line is not needed.
```suggestion
```

File: R/uniformity_test.R [hunk 3]
```diff
@@ -185,13 +189,8 @@ uniformity_test <- function(pit, test, truncate = NULL) {
...
 .cauchy_combination_test <- function(x, truncate = NULL) {
-  if (truncate) {
-    idx <- which(x < 0.5)
-    if (length(idx) == 0) {
-      stop("Cannot compute truncated Cauchy combination test. ",
-           "No p-values below 0.5 found.")
-    }
-    1 - pcauchy(mean(-qcauchy(x[idx])))
+  if (truncate){
+    1 - pcauchy( mean(ifelse(x < 0.5,-qcauchy(x), 0.0)) )
```
Comment (florence-bockting) [inline R/uniformity_test.R]: Again, just some comments regarding style: 

- Try to avoid white space between opening bracket ( and the next symbol/char/...
- Use a white space between condition and opening curly bracket { for if-loops and other conditionals

```suggestion
  if (truncate) {
    1 - pcauchy(mean(ifelse(x < 0.5, -qcauchy(x), 0.0))) 
```

File: tests/testthat/test-uniformity_test.R [hunk 4]
```diff
@@ -38,17 +38,21 @@ test_that("uniformity_test errors on invalid test", {
...
+     for( alpha in c( 0.01, 0.02) ){
+    pvals <- replicate(10000, {
+    x <- runif(200)   # generate under H0
+    uniformity_test(x, test)$pvalue })
+    res <- mean(pvals<alpha)
+
+    expect_equal(res, alpha, tolerance =1e-2 )
+     }
+    }
 })
```
Comment (florence-bockting) [inline tests/testthat/test-uniformity_test.R]: For this test there are several comments:

- one is again to style (I mentioned the aspect of white spaces already earlier), here we see also problems with the indentation. Make sure that the indentation after for-loops / conditionals / etc. is correct.
- then regarding the position of the for-loop `for (alpha in c(0.01, 0.02)) {`: is it necessary that the replications are redone for every alpha?
- as this test introduces a random component, we provide seeds (`set.seed()`)
- this is a very extensive test (takes quite long). So we don't want it on cran therefore I added `skip_on_cran` (tagging here also @paul-buerkner whether that's okay like this or whether you want to skip the test somehow entirely?)

```suggestion
test_that("uniformity_test is calibrated", {
  skip_on_cran()
  set.seed(4711)
  nsim <- 10000
  for (test in c("POT", "PIET", "PRIT")) {
    pvals <- replicate(nsim, uniformity_test(runif(200), test)$pvalue)
    for (alpha in c(0.01, 0.02)) {
      res <- mean(pvals < alpha)
      expect_equal(res, alpha, tolerance = 1e-2)
    }
  }
})
```

File: tests/testthat/test-uniformity_test.R [hunk 5]
```diff
@@ -38,17 +38,21 @@ test_that("uniformity_test errors on invalid test", {
...
 
+
```
Comment (florence-bockting) [inline tests/testthat/test-uniformity_test.R]: I think this extra line is not needed.

```suggestion
```

File: tests/testthat/test-uniformity_test.R [hunk 6]
```diff
@@ -186,7 +190,7 @@ test_that("cauchy_combination_test handles truncate = TRUE", {
 
   x <- c(0.1, 0.2, 0.3, 0.4, 0.7, 0.8)
   result <- .cauchy_combination_test(x, truncate = TRUE)
-  expected <- 1 - pcauchy(mean(-qcauchy(c(0.1, 0.2, 0.3, 0.4))))
+  expected <- 1 - pcauchy(mean(-qcauchy(x)*(x<0.5)))
```
Comment (florence-bockting) [inline tests/testthat/test-uniformity_test.R]: Just minor adjustment to linting again:

```suggestion
  expected <- 1 - pcauchy(mean(-qcauchy(x) * (x < 0.5)))
```

File: tests/testthat/test-uniformity_test.R [hunk 7]
```diff
@@ -202,7 +206,9 @@ test_that("cauchy_combination_test handles boundary values", {
   expect_true(is.nan(.cauchy_combination_test(c(0, 1), truncate = FALSE)))
   # TODO: if 1 included in vector, CCT will always evaluate to 0
   # as the mean evaluates to Inf and 1 - cdf(Inf) = 1 - 1 = 0
-  expect_equal(.cauchy_combination_test(c(0, 0.3, 0.4, 1), truncate = TRUE), 0)
+  x2 <- c(0, 0.3, 0.4, 1)
+  expect_equal(.cauchy_combination_test(x2, truncate = TRUE), 1 - pcauchy(mean(ifelse(x2<0.5,-qcauchy(x2),0))) )
+
```
Comment (florence-bockting) [inline tests/testthat/test-uniformity_test.R]: The lines goes over the 80 character rule which we try to adhere. We try to break long lines into multiple lines. The following change shows an example:

```suggestion
  x2 <- c(0, 0.3, 0.4, 1)
  expect_equal(
    .cauchy_combination_test(x2, truncate = TRUE),
    1 - pcauchy(mean(ifelse(x2 < 0.5, -qcauchy(x2), 0)))
  )
```

File: tests/testthat/test-uniformity_test.R [hunk 8]
```diff
@@ -216,4 +222,4 @@ test_that("compute_cauchy computes correct transformations", {
 
   # For x = 0.75: tan((0.5 - 0.75) * pi) = tan(-0.25 * pi) = -1
   expect_equal(.compute_cauchy(0.75), -1, tolerance = 1e-10)
-})
\ No newline at end of file
+})
```
Comment (florence-bockting) [inline tests/testthat/test-uniformity_test.R:261]: Minor thing (I like to forget myself) is an empty line at the end of the file.

```suggestion
})

```

**Review by florence-bockting — CHANGES_REQUESTED** [review]
Thank you very much for the PR @hermanFTT.  It looks very good, I have only minor changes which are mainly styling aspects (indentation, white spaces, new lines, etc.)

Otherwise it looks very good and I think after this small changes it is good to go from my side.

Tagging @paul-buerkner for having a short second look at it (specifically regarding the calibration test)

Thank you!

**Comment by florence-bockting** [comment]
It would be nice if you @hermanFTT could additionally provide a short information about your change in the `NEWS.md`
Simply add a new section like

```
# posterior (development)
### Bug Fixes

* description of your change by @hermanFTT (#462)

# posterior 1.7.1
...
```

**Comment by paul-buerkner** [comment]
I am a bit out of the code for the uniformity tests right now. @avehtari can you check? I assume this PR was pre-discussed with you?

File: tests/testthat/test-uniformity_test.R [hunk 9]
```diff
@@ -2,7 +2,7 @@
 set.seed(123)
 
 test_that("uniformity_test returns list with pvalue and pointwise", {
-  pit <- runif(20)
+  pit <- runif(100)
   for (test in c("POT", "PIET", "PRIT")) {
```
Comment (avehtari) [inline tests/testthat/test-uniformity_test.R:6]: Should PRIT be tested with discrete pit? Now pit is continuous from runif

**Review by avehtari — CHANGES_REQUESTED** [review]
- I agree on skipping slow test on CRAN
- maybe add tests for the new no-`p<0.5` → `0.5` behavior and for `truncate=FALSE` to improve coverage?

**Comment by florence-bockting** [comment]
> * I agree on skipping slow test on CRAN
> * maybe add tests for the new no-`p<0.5` → `0.5` behavior and for `truncate=FALSE` to improve coverage?

Thank you for your review!

@hermanFTT do you want to give it a try and suggest follow-up tests?
Regarding the suggested code changes: The way how this usually works, is that you go over the suggestions, and when you are "fine" with them, then you click "Apply suggestion". A drop-down window opens and you can provide a text which will be used as commit message once you click "commit changes". This will automatically commit to the branch. You can then locally on your computer "pull" the changes and then you will directly see the code changes.

Regarding the order, I would suggest, 
1. you go first over my code suggestions and decide for each whether you want to apply the change or not.
2. Then you change to your local setup and use "git pull" to update your local branch.
3. Then you write the new tests (according to Aki's suggestion) and edit the NEWS.md.
4. Then git add, commit, and push the changes and tag (@florence-bockting) me again.

When you are uncertain or have any question in one of these steps, don't hesitate and let me know.

**Comment by hermanFTT** [comment]
@florence-bockting I have reviewed and updated the PR based on the review/suggestions. See, e.g,  additional unit test for the case `truncated = TRUE` when all pointwise $p$-values to be combined are above 0.5 .

**Comment by florence-bockting** [comment]
> @florence-bockting I have reviewed and updated the PR based on the review/suggestions. See, e.g, additional unit test for the case `truncated = TRUE` when all pointwise p -values to be combined are above 0.5 .

Perfect, looks very good. Thank you @hermanFTT.
What do you think about adding the following two tests which address explicitly the point of using discrete values for PRIT? I think, then we can merge the PR.

```
test_that("PRIT is calibrated for discrete pit values", {
  skip_on_cran()
  set.seed(4711)
  nsim <- 1000

  draw_pit <- function() pbinom(rbinom(100, 5, 0.5), 5, 0.5)
  pvals <- replicate(nsim, uniformity_test(draw_pit(), "PRIT")$pvalue)
  res <- mean(pvals < 0.01)
  expect_equal(res, 0.01, tolerance = 1e-2)
})

test_that("prit_test gives equal p-values to tied PIT values", {
  # Let n = 4, x = c(0.25, 0.25, 0.75, 0.75)
  # scaled_ecdf = 4 * c(0.5, 0.5, 1, 1) = c(2, 2, 4, 4)
  # For x = 0.25: pbinom(2, 4, 0.25) = 0.94921875
  #               dbinom(2, 4, 0.25) = 0.2109375
  #               probs1 = 0.94921875 - 0.10546875 = 0.84375
  #               p_val = 2 * (1 - 0.84375) = 0.3125
  # For x = 0.75: pbinom(4, 4, 0.75) = 1
  #               dbinom(4, 4, 0.75) = 0.31640625
  #               probs1 = 1 - 0.158203125 = 0.841796875
  #               p_val = 2 * 0.158203125 = 0.31640625

  x <- c(0.25, 0.25, 0.75, 0.75)
  expected <- c(0.3125, 0.3125, 0.31640625, 0.31640625)

  expect_equal(.prit_test(x), expected, tolerance = 1e-10)
})
```

**Comment by hermanFTT** [comment]
I actually had similar tests but somehow forgot to commit and push. It is now updated.

File: tests/testthat/test-uniformity_test.R (hunk 9 above)
Comment (florence-bockting) [reply tests/testthat/test-uniformity_test.R:6]: Done with Commit: https://github.com/stan-dev/posterior/pull/462/commits/2d89594c04b3cd25149adb55cd536bd3c5d98856

**Review by florence-bockting — APPROVED** [review]

**Comment by paul-buerkner** [comment]
Thanks! @florence-bockting you can then merge whenever you feel like (I assume you have the rights?)
